### PR TITLE
Context fields

### PR DESCRIPTION
**Pull request by pzabolotniy**
@hhkbp2 , i'd like to log any fields, not only declared in https://github.com/hhkbp2/go-logging/blob/master/formatter.go#L15
So, i suggest to extend formatter to use any user-defined fields.
Value for such fields must be set in application.
here is my PR
what do you think?

### CODE REVIEW

**Comment by pzabolotniy** [comment]
hi, @hhkbp2 !
So, what do you think about this feature?

**Comment by hhkbp2** [comment]
Hi @pzabolotniy ,

Seeing from the codes, the context fields are added as an alias way to write the format message, arguments. Most of the case, the printf format arguments and some helper function will be good enough for usage.

In term of coupling, I am not sure whether it's good idea to bind the context fields to the logger. It adds extra and implicit work to handle(initialize, copy) the context fields for anyone who want to define a new logger(for whatever reason), which changes the original clean design. A good idea would be adding a context in the interface, like

`
Infof(context, "format with context field", args...)
`

but that is not much different to

`
Infof("format with hard code field", args with field value...)
`

As for performance, since `context fields` is a map, you should copy it every time for

`
1. the logger initialization
2. record initialization
`

so that it has a clear semantics that it would not affect the behavior even if the logger initialization argument map change, or the logger internal map changes. But then it becomes a performance burden to clone the map all around.

And since the field name is determined in runtime, the code can't generate a bunch of fixed functions to extract record field in order. It has to do the map query for every record if field exists, which will reduce the performance.

In summary, I think this PR should be revisited for both design and performance. We would consider merging it into master branch, after we have a clean design and a test which proves it will not impact heavily on the performance.

**Comment by pzabolotniy** [comment]
We use context in our http rest api application. We set uniq context ID for every request, and then use contexted-Logger all over of application. While, no other packages know about context ID, they just log their work. So, "printf"-solution can not be used here.

The solution i've implemented has been dublicated from logrus idea: it also makes a copy of logger when initiliaze context fields.

@hhkbp2 , could you share sources of the benchmark tests, so i could test the performance?

**Comment by hhkbp2** [comment]
@pzabolotniy 

I took a look at the `logrus` code and see that its logger has different design(much simpler than `logger` in this project). In `go-logging`(also python `logging`), usually a standard logger is created by calling the module interface `GetLogger()`, which will construct and registered its name to the Manager so that it could be re-used. It's more like a global logger with registered name(that's why it has namespace hierarchy).

In this case, IMHO it should pass around and use the context ID when logging, instead of passing around and using the contexted-logger all over the codes. If you insist to passing around the logger, you could define a customized logger(deriving from the standard logger to reuse codes). In this customized logger, you could simply wrap a call like

`Infof("Message", args...)`

to actually calling

`Infof("[CtxID=%d] Message", ctxID, args...)`

and you need to modify `findCaller()` to get the correct caller info, then don't use `GetLogger()` to construct the customized logger.

It's a better solution without the defects on design, and it seems no big impact on performance. (Just one more function call down the calling stack and one more step on the `findCaller` loop. In this case, the new customized logger code would be an extension of current module, we should not merge it into this module. To be sure about performance, you could make a benchmark test(logging into a file in a same setting) and do a third-party comparison using customized logger, contexted field record(this PR), and the current master branch.

**Comment by pzabolotniy** [comment]
hi, @hhkbp2 !
I understood the difference between our points of view. I didn't take in attention the `namespace hierarchy`. We do not use it in our application. We call `GetLogger()` once, when app starts.

I'll think about an extension of  current module.

Solution
`Infof("[CtxID=%d] Message", ctxID, args...)`
has disadvantages
1. message format can not be configured in config file.
2. there is no way to effect on caller's depth search. It is hard coded to 3.

So, you may close this PR. Thanks!

**Comment by hhkbp2** [comment]
@pzabolotniy 

Using the wrapping way I mentioned above, it's able to re-write the `findCaller()` function to go any depth in the calling stack as you need. And for the formatter format, it could be configured in config file as before. No config access to the new wrapping logger and wrapper, but for sure you could extend current codes to do that.

File: logger.go [hunk 1]
```diff
@@ -183,6 +188,16 @@ func findCaller() *CallerInfo {
 	}
 }
 
+type CtxFields map[string]string
+
+func (fields *CtxFields) save(k, v string) {
```
Comment (ilius) [inline logger.go]: Does the receiver need to be a pointer? Maps are already pointers.
And any reason not use `set`? (as opposed to `get`)

File: record.go [hunk 2]
```diff
@@ -26,6 +26,8 @@ type LogRecord struct {
 	UseFormat   bool
 	Args        []interface{}
 	Message     string
+	CtxID       string
```
Comment (ilius) [inline record.go]: `CtxID` does not seem to be used. What is it for?

File: formatter.go [hunk 3]
```diff
@@ -11,40 +11,47 @@ import (
 // the attribute string.
 type ExtractAttr func(record *LogRecord) string
 
+type RecordValueReceiver func(record *LogRecord) interface{}
+
+func GetValueForField(fieldName string, record *LogRecord) interface{} {
+	var value interface{}
```
Comment (ilius) [inline formatter.go]: We don't need `value` variable. Simpler to just `return` in each `case`

**Comment by hhkbp2** [comment]
@ilius I'm not sure @pzabolotniy is still working on this. But it seems not gonna get into the master. You could continue your discussion util I close it later.

**Comment by hhkbp2** [comment]
Will close. Reopen it if necessary.
